File: backend/src/schemas/research.py
"""研究图谱相关 Pydantic 模型。"""

from datetime import datetime
from typing import Any, Optional

from pydantic import BaseModel


class ResearchTopicCreate(BaseModel):
    title: str
    description: Optional[str] = None


class ResearchTopicUpdate(BaseModel):
    title: Optional[str] = None
    description: Optional[str] = None
    status: Optional[str] = None
    summary: Optional[str] = None


class ResearchTopicRead(BaseModel):
    id: int
    title: str
    description: Optional[str] = None
    status: str
    summary: Optional[str] = None
    last_checked_at: Optional[datetime] = None
    source_count: int = 0
    claim_count: int = 0
    conflict_count: int = 0
    entity_count: int = 0
    created_at: datetime
    updated_at: datetime


class ResearchSourceRead(BaseModel):
    id: int
    topic_id: int
    title: str
    url: Optional[str] = None
    canonical_url: Optional[str] = None
    publisher: Optional[str] = None
    source_type: str
    published_at: Optional[datetime] = None
    fetched_at: Optional[datetime] = None
    trust_level: str
    status: str
    raw_excerpt: Optional[str] = None
    metadata_json: Optional[dict[str, Any]] = None
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class ResearchEvidenceRead(BaseModel):
    id: int
    topic_id: int
    source_id: Optional[int] = None
    quote: str
    location: Optional[str] = None
    kind: str
    metadata_json: Optional[dict[str, Any]] = None
    created_at: datetime

    model_config = {"from_attributes": True}


class ResearchClaimCreate(BaseModel):
    claim_text: str
    status: str = "pending"
    confidence: int = 0
    claim_type: Optional[str] = None
    adopted: bool = False
    reasoning: Optional[str] = None
    evidence_ids: list[int] = []
    entity_ids: list[int] = []
    entity_names: list[str] = []


class ResearchClaimUpdate(BaseModel):
    status: Optional[str] = None
    confidence: Optional[int] = None
    adopted: Optional[bool] = None
    reasoning: Optional[str] = None
    evidence_ids: Optional[list[int]] = None


class ResearchConflictResolve(BaseModel):
    accepted_claim_id: int
    rejected_claim_id: int


class ResearchClaimRead(BaseModel):
    id: int
    topic_id: int
    claim_text: str
    status: str
    confidence: int
    claim_type: Optional[str] = None
    adopted: bool
    reasoning: Optional[str] = None
    entity_ids: list[int] = []
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class ResearchEntityCreate(BaseModel):
    name: str
    entity_type: Optional[str] = "other"
    description: Optional[str] = None
    confidence: int = 0
    status: str = "active"
    aliases: list[str] = []


class ResearchEntityUpdate(BaseModel):
    name: Optional[str] = None
    entity_type: Optional[str] = None
    description: Optional[str] = None
    confidence: Optional[int] = None
    status: Optional[str] = None
    aliases: Optional[list[str]] = None


class ResearchEntityRead(BaseModel):
    id: int
    topic_id: int
    name: str
    entity_type: Optional[str] = None
    description: Optional[str] = None
    confidence: int = 0
    status: str = "active"
    aliases_json: Optional[list[str]] = None
    created_at: datetime
    updated_at: Optional[datetime] = None

    model_config = {"from_attributes": True}


class ResearchClaimEntityLinkRead(BaseModel):
    id: int
    claim_id: int
    entity_id: int
    topic_id: int
    role: Optional[str] = None
    created_at: datetime

    model_config = {"from_attributes": True}


class ResearchRelationRead(BaseModel):
    id: int
    topic_id: int
    from_type: str
    from_id: int
    to_type: str
    to_id: int
    relation_type: str
    metadata_json: Optional[dict[str, Any]] = None
    created_at: datetime

    model_config = {"from_attributes": True}


class ResearchProposalRead(BaseModel):
    id: int
    topic_id: int
    proposal_type: str
    title: str
    description: Optional[str] = None
    payload_json: Optional[dict[str, Any]] = None
    status: str
    created_at: datetime
    reviewed_at: Optional[datetime] = None
    applied_at: Optional[datetime] = None

    model_config = {"from_attributes": True}


class ResearchProposalUpdate(BaseModel):
    status: Optional[str] = None  # approved | rejected | applied | pending


class ResearchRunRead(BaseModel):
    id: int
    topic_id: int
    status: str
    progress: dict[str, Any]
    error_message: Optional[str] = None
    started_at: Optional[datetime] = None
    finished_at: Optional[datetime] = None
    created_at: datetime
    updated_at: datetime


class ResearchTopicDetail(ResearchTopicRead):
    sources: list[ResearchSourceRead] = []
    evidence: list[ResearchEvidenceRead] = []
    claims: list[ResearchClaimRead] = []
    entities: list[ResearchEntityRead] = []
    relations: list[ResearchRelationRead] = []
    claim_entity_links: list[ResearchClaimEntityLinkRead] = []
    proposals: list[ResearchProposalRead] = []


class BlogPostResearchAttach(BaseModel):
    topic_id: int


class BlogPostClaimAttach(BaseModel):
    claim_id: int
    usage_note: str = ""


class BlogPostResearchLinkRead(BaseModel):
    id: int
    post_id: int
    topic_id: int
    snapshot: dict[str, Any]
    created_at: datetime


class BlogPostClaimLinkRead(BaseModel):
    id: int
    post_id: int
    claim_id: int
    topic_id: int
    usage_note: Optional[str] = None
    created_at: datetime


class BlogResearchSummary(BaseModel):
    post_id: int
    topics: list[dict[str, Any]]
    claims: list[dict[str, Any]]

File: frontend/src/api/research.ts
import { API_BASE, apiFetch, readErrorDetail } from "./client";

export interface ResearchTopicSummary {
  id: number;
  title: string;
  description?: string | null;
  status: string;
  summary?: string | null;
  last_checked_at?: string | null;
  source_count: number;
  claim_count: number;
  entity_count: number;
  conflict_count: number;
  created_at: string;
  updated_at: string;
}

export interface ResearchSource {
  id: number;
  topic_id: number;
  title: string;
  url?: string | null;
  canonical_url?: string | null;
  publisher?: string | null;
  source_type: string;
  published_at?: string | null;
  fetched_at?: string | null;
  trust_level: string;
  status: string;
  raw_excerpt?: string | null;
  metadata_json?: Record<string, unknown> | null;
  created_at: string;
  updated_at: string;
}

export interface ResearchEvidence {
  id: number;
  topic_id: number;
  source_id?: number | null;
  quote: string;
  location?: string | null;
  kind: string;
  metadata_json?: Record<string, unknown> | null;
  created_at: string;
}

export interface ResearchClaim {
  id: number;
  topic_id: number;
  claim_text: string;
  status: string;
  confidence: number;
  claim_type?: string | null;
  adopted: boolean;
  reasoning?: string | null;
  entity_ids: number[];
  created_at: string;
  updated_at: string;
}

export interface ResearchEntity {
  id: number;
  topic_id: number;
  name: string;
  entity_type?: string | null;
  description?: string | null;
  confidence: number;
  status: string;
  aliases_json?: string[] | null;
  created_at: string;
  updated_at?: string | null;
}

export interface ResearchClaimEntityLink {
  id: number;
  claim_id: number;
  entity_id: number;
  topic_id: number;
  role?: string | null;
  created_at: string;
}

export interface ResearchRelation {
  id: number;
  topic_id: number;
  from_type: string;
  from_id: number;
  to_type: string;
  to_id: number;
  relation_type: string;
  metadata_json?: Record<string, unknown> | null;
  created_at: string;
}

export interface ResearchConflictResolvePayload {
  accepted_claim_id: number;
  rejected_claim_id: number;
}

export interface ResearchProposal {
  id: number;
  topic_id: number;
  proposal_type: string;
  title: string;
  description?: string | null;
  payload_json?: Record<string, unknown> | null;
  status: string;
  created_at: string;
  reviewed_at?: string | null;
  applied_at?: string | null;
}

export interface ResearchRun {
  id: number;
  topic_id: number;
  status: string;
  progress: Record<string, unknown>;
  error_message?: string | null;
  started_at?: string | null;
  finished_at?: string | null;
  created_at: string;
  updated_at: string;
}

export interface ResearchTopicDetail extends ResearchTopicSummary {
  sources: ResearchSource[];
  evidence: ResearchEvidence[];
  claims: ResearchClaim[];
  entities: ResearchEntity[];
  relations: ResearchRelation[];
  claim_entity_links: ResearchClaimEntityLink[];
  proposals: ResearchProposal[];
}

export interface BlogPostResearchLink {
  id: number;
  post_id: number;
  topic_id: number;
  snapshot: Record<string, unknown>;
  created_at: string;
}

export interface BlogResearchSummary {
  post_id: number;
  topics: Array<Record<string, unknown>>;
  claims: Array<Record<string, unknown>>;
}

export interface ResearchDraftReference {
  evidence_id?: number;
  quote?: string;
  source_id?: number;
  title?: string;
  url?: string;
}

export interface ResearchDraftPreview {
  title: string;
  outline: string[];
  content: string;
  references: ResearchDraftReference[];
}

export async function listResearchTopics(): Promise<ResearchTopicSummary[]> {
  const res = await apiFetch(`${API_BASE}/research/topics`);
  if (!res.ok) throw new Error("Failed to fetch research topics");
  return res.json();
}

export async function createResearchTopic(data: {
  title: string;
  description?: string | null;
}): Promise<ResearchTopicSummary> {
  const res = await apiFetch(`${API_BASE}/research/topics`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(data),
  });
  if (!res.ok) {
    const err = await readErrorDetail(res, "创建研究主题失败");
    throw new Error(err);
  }
  return res.json();
}

export async function getResearchTopic(topicId: number): Promise<ResearchTopicDetail> {
  const res = await apiFetch(`${API_BASE}/research/topics/${topicId}`);
  if (!res.ok) throw new Error("Failed to fetch research topic");
  return res.json();
}

export async function deleteResearchTopic(topicId: number): Promise<void> {
  const res = await apiFetch(`${API_BASE}/research/topics/${topicId}`, { method: "DELETE" });
  if (!res.ok) {
    const err = await readErrorDetail(res, "删除研究主题失败");
    throw new Error(err);
  }
}

export async function runResearchTopic(topicId: number, idempotencyKey?: string): Promise<ResearchRun> {
  const headers = new Headers();
  if (idempotencyKey) headers.set("Idempotency-Key", idempotencyKey);
  const res = await apiFetch(`${API_BASE}/research/topics/${topicId}/run`, {
    method: "POST",
    headers,
  });
  if (!res.ok) throw new Error("Failed to run research topic");
  return res.json();
}

export async function listResearchRuns(topicId: number): Promise<ResearchRun[]> {
  const res = await apiFetch(`${API_BASE}/research/topics/${topicId}/runs`);
  if (!res.ok) throw new Error("Failed to list research runs");
  return res.json();
}

export async function draftResearch(topicId: number): Promise<ResearchDraftPreview> {
  const res = await apiFetch(`${API_BASE}/research/topics/${topicId}/draft-preview`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
  });
  if (!res.ok) throw new Error("Failed to generate research draft preview");
  return res.json();
}

export async function updateResearchClaim(claimId: number, data: {
  status?: string;
  confidence?: number;
  adopted?: boolean;
  reasoning?: string | null;
  evidence_ids?: number[];
  entity_ids?: number[];
  entity_names?: string[];
}): Promise<ResearchClaim> {
  const res = await apiFetch(`${API_BASE}/research/claims/${claimId}`, {
    method: "PATCH",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(data),
  });
  if (!res.ok) {
    const err = await readErrorDetail(res, "更新事实声明失败");
    throw new Error(err);
  }
  return res.json();
}

export async function resolveResearchConflict(
  relationId: number,
  payload: ResearchConflictResolvePayload,
): Promise<ResearchRelation> {
  const res = await apiFetch(`${API_BASE}/research/conflicts/${relationId}/resolve`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(payload),
  });
  if (!res.ok) {
    const err = await readErrorDetail(res, "冲突解决失败");
    throw new Error(err);
  }
  return res.json();
}

export async function updateResearchProposal(proposalId: number, data: {
  status?: string;
}): Promise<ResearchProposal> {
  const res = await apiFetch(`${API_BASE}/research/proposals/${proposalId}`, {
    method: "PATCH",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(data),
  });
  if (!res.ok) {
    const err = await readErrorDetail(res, "更新提案失败");
    throw new Error(err);
  }
  return res.json();
}

export async function attachResearchTopicToPost(postId: number, topicId: number): Promise<BlogPostResearchLink> {
  const res = await apiFetch(`${API_BASE}/blog/posts/${postId}/research-topics`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ topic_id: topicId }),
  });
  if (!res.ok) throw new Error("Failed to attach research topic");
  return res.json();
}

export async function getBlogResearchSummary(postId: number): Promise<BlogResearchSummary> {
  const res = await apiFetch(`${API_BASE}/blog/posts/${postId}/research-summary`);
  if (!res.ok) throw new Error("Failed to fetch blog research summary");
  return res.json();
}

